# Algoritmo KNN usando a linguagem de programação Python
### Nesse arquivo será feito a análise de uma base de dados para risco de crédito usando as ideias do algoritmo KNN.

#### Pré-processamento dos dados

In [3]:
# importando a biblioteca pandas do python
import pandas as pd

# importando a base de dados para o objeto "dataframe"
dataframe = pd.read_csv('Dados de Credito.csv', encoding = 'utf-8', sep = ',')

# substituindo valores inconsistentes no campo idades pela média das idades consistentes
dataframe.loc[dataframe.age < 0, 'age'] = 40.92

# separando os atributos previsores do meta classe
previsores = dataframe.iloc[:, 1:4].values
classe = dataframe.iloc[:, 4].values

# importando a biblioteca sklearn do python
from sklearn.impute import SimpleImputer
# função "SimpleImputer" responsável por corrigir valores faltantes na base de dados

# importando a biblioteca numpy do python
import numpy as np

# criando o objeto "imputer"
imputer = SimpleImputer(missing_values = np.nan, strategy = "mean")

# fazendo o treinamento com a base de dados para correção de valores faltantes
imputer = imputer.fit(previsores[:,0:3])

# corrigindo os valores faltantes usando o objeto criado "imputer"
previsores[:, 0:3] = imputer.transform(previsores[:, 0:3])

# importando a biblioteca sklearn do python
from sklearn.preprocessing import StandardScaler
# a função "StandardScaler" tem a função de escalonas toda a base de dados para corrigir a discrepância
# dos valores

# criando o objeto "scaler"
scaler = StandardScaler()

# escalonando todos os atributos para auxiliar no cálculo de distâncias euclidianas
previsores = scaler.fit_transform(previsores)

# importando a biblioteca sklearn do python
from sklearn.model_selection import train_test_split
# a função "train_test_split" tem a importância de separar modelos de treinamento e modelos de teste
# em uma base de dados

previsores_treinamento, previsores_teste, classe_treinamento, classe_teste = train_test_split(previsores,
                                                                                             classe,
                                                                                             test_size = 0.25,
                                                                                             random_state = 0)

#### Algoritmo KNN com Python

In [5]:
# importando a biblioteca sklearn do python
from sklearn.neighbors import KNeighborsClassifier
# função "KNeighborsClassifier" responsável por fazer a classificação dos atributos pelo método do algoritmo
# KNN

In [6]:
# criando o objeto "classificador"
classificador = KNeighborsClassifier(n_neighbors = 5, metric = "minkowski", p = 2)

In [7]:
# fazendo o treinamento com a base de dados usando o objeto "classificador"
classificador.fit(previsores_treinamento, classe_treinamento)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [9]:
# aplicando a transformação na base de dados usando o objeto "classificador"
previsoes = classificador.predict(previsores_teste)

In [10]:
# observando a base de dados obtida pelo algoritmo
previsoes

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [12]:
# observando a base de dados original reservada para o teste
classe_teste

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,

In [13]:
# importando a biblioteca sklearn do python
from sklearn.metrics import confusion_matrix, accuracy_score

In [14]:
# calculando a porcentagem de acertos do algoritmo
precisao = accuracy_score(classe_teste, previsoes)

In [15]:
# visualizando a precisão de acertos do algoritmo
precisao

0.986

A precisão de acertos do algoritmo em acertar as classificações foi de **98.6%**.

In [16]:
# criando a matriz de confusão para analisar a quantidade de acertos e erros
matriz = confusion_matrix(classe_teste, previsoes)

In [17]:
# visualizando a matriz de confusão
matriz

array([[432,   4],
       [  3,  61]], dtype=int64)

Analisando a matriz de confusão acima, é possível perceber que 432 dados referentes as pessoas que não pagariam o que devem foi classificado corretamente e 61 dados referentes as pesssoas que pagariam o que devem foi classificado corretamente. Entretanto, 4 dados referentes a não pagadores foram classificados como pagadadores e 3 dados de pessoas pagadoras foram classificadas como não pagadoras. Portanto, a quantidade de acertos foi 493/500 ou 98.6%.

In [18]:
# importando a biblioteca collections
import collections

In [19]:
# visualizando a quantidade de registros para cada uma das classes
collections.Counter(classe_teste)

Counter({1: 64, 0: 436})

### Alguma Dúvida? Entre em Contato Comigo:

- [Me envie um e-mail](mailto:alyssonmachado388@gmail.com);